### PR TITLE
Fix buildah issue of using `overlay` as storage driver

### PR DESCRIPTION
**Pull request by Divyansh42**
Signed-off-by: divyansh42 <diagrawa@redhat.com>

### Description
- This PR adds `mount program` which uses `fuse-overlayfs` to solve the issue https://github.com/redhat-actions/buildah-build/issues/45

### Related Issue(s)
Close: #45

### Checklist

- [ ] This PR includes a documentation change
- [x] This PR does not need a documentation change
---
- [ ] This PR includes test changes
- [x] This PR's changes are already tested
---
- [ ] This change is a patch change
- [x] This change is a minor change
- [ ] This change is a major (breaking) change

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -34,6 +36,13 @@ export async function run(): Promise<void> {
     // remove white spaces (if any) in archs input
     archs = archs.replace(/\s+/g, "");
 
+    // To solve https://github.com/redhat-actions/buildah-build/issues/45
+    const homedir = os.homedir();
+    const storageFilePaths: string[] = [ path.join(homedir, ".config/containers/storage.conf"),
```
Comment (tetchel) [inline src/index.ts:41]: ~/.config usually comes from `XDG_CONFIG_HOME`. Use that if it's set, or default to ~/.config

https://wiki.archlinux.org/index.php/XDG_Base_Directory#User_directories

File: src/index.ts [hunk 2]
```diff
@@ -34,6 +36,13 @@ export async function run(): Promise<void> {
...
+        "/etc/containers/storage.conf" ];
```
Comment (tetchel) [inline src/index.ts:42]: ] on new line

File: src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        core.info("Reading storage file to find out the storage driver used");
+        const fileContent = ini.parse(fs.readFileSync(filePath, "utf-8"));
+        const storageDriver = fileContent.storage.driver;
+        core.info(`Storage driver is set to "${storageDriver}"`);
+
+        // Check for storage-driver, if overlay then and
+        // add mount program as "fuse-overlayfs" in ~/.config/containers/storage.conf
+        if (storageDriver === "overlay") {
+            // Backup file if ~/.config/containers/storage.conf exists
+            if (filePath === filePaths[0]) {
```
Comment (tetchel) [inline src/utils.ts:19]: Since the logic for [0] and [1] (home dir vs /etc/) is fairly different, maybe it should be two different function calls or functions because this is hard to follow

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            core.debug(`Storage file exists at ${filePath}`);
+            return filePath;
+        }
+    }
+    return "";
+}
+
+function backupFile(filePath: string): void {
+    core.debug(`Creating backup of the existing "storage.conf" file`);
+    fs.copyFileSync(`${filePath}`, `${filePath}.backup`);
```
Comment (tetchel) [inline src/utils.ts:43]: put .backup in a constant

or
```
function getBackupFileName(filePath: string): string {
    return filePath + ".backup";
}
```

and use that function everywhere you need it

File: src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    core.debug(`Backup created at ${filePath}.backup`);
+}
+
+function addMountProgram(): void {
+    core.debug(`Creating directory "~/.config/containers" recursively`);
+    const homedir = os.homedir();
+    const targetDir = path.join(homedir, ".config/containers");
```
Comment (tetchel) [inline src/utils.ts:50]: shouldn't this path be passed down by the caller (index.ts)?

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    fs.mkdirSync(targetDir, { recursive: true });
+    core.debug("Directory created successfully");
+
+    const fileName = "storage.conf";
+
+    const mountProgramData = "[storage.options]\nmount_program=\"/usr/bin/fuse-overlayfs\"";
```
Comment (tetchel) [inline src/utils.ts:57]: can you use the ini library to write this?

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    core.info(`Creating "${fileName}" and adding "fuse-overlayfs" in the mount program`);
+    fs.appendFileSync(`${targetDir}/${fileName}`, mountProgramData);
```
Comment (tetchel) [inline src/utils.ts:60]: this should be a write not an append since we know it doesn't exist. if it did exist and it had any content then this would likely make the file invalid.

**Review by tetchel — CHANGES_REQUESTED** [review]
some comments, also I would prefer if you used the async variants and used await everywhere.

I liike to do `import fs from "fs/promises"` which gives a good promisified api.

File: src/index.ts [hunk 8]
```diff
@@ -44,6 +53,11 @@ export async function run(): Promise<void> {
     if (tagsList.length > 1) {
         await cli.tag(image, tagsList);
     }
+
+    // Restore if original file is changed
+    core.debug("Restoring the original file to avoid any breaking change for future jobs");
+    restoreFile(storageFilePaths[0]);
```
Comment (tetchel) [inline src/index.ts:59]: wrap this in a try/catch with an error log, but the action should not fail if this fails

File: src/index.ts [hunk 9]
```diff
@@ -34,6 +36,13 @@ export async function run(): Promise<void> {
     // remove white spaces (if any) in archs input
     archs = archs.replace(/\s+/g, "");
 
+    // To solve https://github.com/redhat-actions/buildah-build/issues/45
+    const homedir = os.homedir();
+    const storageFilePaths: string[] = [ path.join(homedir, ".config/containers/storage.conf"),
+        "/etc/containers/storage.conf" ];
+    // If overlay is used as storage driver then set "fuse-overlayfs as mount program"
+    checkStorageDriver(storageFilePaths);
```
Comment (tetchel) [inline src/index.ts:44]: wrap this in a try/catch with an error log, but the action should not fail if this fails (though I suppose it may fail at a later step due to the issue we are trying to work around)

File: src/utils.ts (hunk 3 above)
Comment (tetchel) [reply src/utils.ts:19]: Remove the array and have two strings for home/etc locations instead and then refactor these two functions accordingly

it would be necessary for https://github.com/redhat-actions/buildah-build/pull/51/files#r609678313 too

File: src/buildah.ts [hunk 10]
```diff
@@ -25,10 +27,22 @@ interface Buildah {
...
+    public storageOptsEnv = "";
+
     constructor(executable: string) {
         this.executable = executable;
     }
 
+    async checkFuseOverlayfs(): Promise<void> {
+        const fuseOverlayfsPath = await io.which("fuse-overlayfs");
+
+        if (fuseOverlayfsPath.startsWith("/usr/bin")) {
```
Comment (tetchel) [inline src/buildah.ts:39]: why check this? why not use it if it's installed elsewhere?

File: src/buildah.ts [hunk 11]
```diff
@@ -25,10 +27,22 @@ interface Buildah {
...
+            if (await checkStorageDriver()) {
+                this.storageOptsEnv = "overlay.mount_program=/usr/bin/fuse-overlayfs";
```
Comment (tetchel) [inline src/buildah.ts:41]: use the path from `which` here instead of hardcoding it..

File: src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        if (await fileExists(filePath)) {
+            core.debug(`Storage file exists at ${filePath}`);
+            const fileContent = ini.parse(await fs.readFile(filePath, "utf-8"));
+            storageDriver = fileContent.storage.driver;
+        }
+    }
+    return storageDriver;
+}
+
+export async function checkStorageDriver(): Promise<boolean> {
```
Comment (tetchel) [inline src/utils.ts:20]: maybe a more accurate name for this would be `isStorageDriverOverlay`

File: src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+async function findStorageDriver(filePaths: string[]): Promise<string> {
+    let storageDriver = "";
+
+    for (const filePath of filePaths) {
+        core.debug(`Checking if the storage file exists at ${filePath}`);
+        if (await fileExists(filePath)) {
+            core.debug(`Storage file exists at ${filePath}`);
+            const fileContent = ini.parse(await fs.readFile(filePath, "utf-8"));
+            storageDriver = fileContent.storage.driver;
```
Comment (tetchel) [inline src/utils.ts:14]: if storage driver is set in the first file but not the second one, then this function will return "", which is not the correct behaviour

File: src/utils.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    return storageDriver;
+}
+
+export async function checkStorageDriver(): Promise<boolean> {
+    let xdgConfigHome = "~/.config";
+    if (process.env.XDG_CONFIG_HOME) {
+        xdgConfigHome = process.env.XDG_CONFIG_HOME;
+    }
+    const filePaths: string[] = [ "/etc/containers/storage.conf",
```
Comment (tetchel) [inline src/utils.ts:25]: if you are spacing an array over multiple lines,

```
const a = [   // newline
  "item1",
  "item2",  // newline
];

File: src/utils.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            core.debug(`Storage file exists at ${filePath}`);
+            const fileContent = ini.parse(await fs.readFile(filePath, "utf-8"));
+            storageDriver = fileContent.storage.driver;
+        }
+    }
+    return storageDriver;
+}
+
+export async function checkStorageDriver(): Promise<boolean> {
+    let xdgConfigHome = "~/.config";
```
Comment (tetchel) [inline src/utils.ts:21]: use `os.homedir` instead of `~`

**Review by tetchel — CHANGES_REQUESTED** [review]

File: src/buildah.ts [hunk 16]
```diff
@@ -25,10 +27,22 @@ interface Buildah {
...
     private readonly executable: string;
 
+    public storageOptsEnv = "";
+
     constructor(executable: string) {
         this.executable = executable;
     }
 
+    async checkFuseOverlayfs(): Promise<void> {
+        const fuseOverlayfsPath = await io.which("fuse-overlayfs");
```
Comment (tetchel) [inline src/buildah.ts:37]: you have to try/catch this because `io.which` throws if it's not found

```
 If you check and the tool does not exist, it will throw.
```

File: src/utils.ts (hunk 13 above)
Comment (Divyansh42) [reply src/utils.ts:14]: Yeah, let me correct

File: src/buildah.ts [hunk 17]
```diff
@@ -25,10 +26,28 @@ interface Buildah {
...
+    // If yes, add mount program to use "fuse-overlayfs"
+    async setStorageOptsEnv(): Promise<void> {
+        if (await isStorageDriverOverlay()) {
+            const fuseOverlayfsPath = await findFuseOverlayfsPath();
+            if (fuseOverlayfsPath) {
+                core.info(`Overriding storage mount_program with "fuse-overlayfs" in environment`);
+                this.storageOptsEnv = `overlay.mount_program=${fuseOverlayfsPath}`;
+            }
+            else {
+                core.warning(`"fuse-overlayfs" is not found. Install it before running this action`);
```
Comment (tetchel) [inline src/buildah.ts:46]: This shouldn't be a warning; there are use-cases like our self-hosted runenr that use different drivers.

Comment (tetchel) [reply src/buildah.ts:46]: `core.info("Storage driver is not 'overlay', so not overriding storage configuration")`

File: src/utils.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+export async function findFuseOverlayfsPath(): Promise<string> {
+    let fuseOverlayfsPath = "";
+    try {
+        fuseOverlayfsPath = await io.which("fuse-overlayfs");
+    }
+    catch (err) {
+        core.debug(err);
+    }
+
+    return fuseOverlayfsPath;
```
Comment (tetchel) [inline src/utils.ts:55]: the emptry string return could cause a problem here. I know that you check if the string is non-zero length at the call site but this is not a good design to rely on the caller doing that check.

This function should return `string | undefined` and `undefined` if fuse-overlayfs is not installed. so the caller is forced to check for a valid return

**Review by tetchel — CHANGES_REQUESTED** [review]

File: src/buildah.ts [hunk 19]
```diff
@@ -25,10 +26,32 @@ interface Buildah {
...
+    // If yes, add mount program to use "fuse-overlayfs"
+    async setStorageOptsEnv(): Promise<void> {
+        if (await isStorageDriverOverlay()) {
+            const fuseOverlayfsPath = await findFuseOverlayfsPath();
+            if (fuseOverlayfsPath) {
+                core.info(`Overriding storage mount_program with "fuse-overlayfs" in environment`);
+                this.storageOptsEnv = `overlay.mount_program=${fuseOverlayfsPath}`;
+            }
+            else {
+                core.warning(`"fuse-overlayfs" is not found. Install it before running this action.`
```
Comment (tetchel) [inline src/buildah.ts:46]: Need a space between the two parts of this string

**Review by tetchel — APPROVED** [review]
